src/cajal/run_gw.py

```diff
@@ -416,6 +416,108 @@ def gw_pairwise_parallel(
+    :param num_processes: How many Python processes to run in parallel for the computation.
+    :param names: A list of unique cell identifiers for both queries and targets where the name
+        for the i-th query is names[i] and the name for the k-th target is names[len(queries) + k].
+        Or in other words: names is expect to be query names + target names (in that order).
+        This argument is required if gw_dist_csv is not None, or if gw_coupling_mat_csv is
+        not None, and is ignored otherwise.
+    :param gw_dist_csv: If this field is a string giving a file path, the GW distances
+        will be written to this file. A list of cell names must be supplied.
+    :param gw_coupling_mat_csv: If this field is a string giving a file path,
+        the GW coupling
+        matrices will be written to this file. A list of cell names must be supplied.
+    :param return_coupling_mats: Whether the function should return the coupling matrices.
+        Please be warned that for a large
+        number of cells, `couplings` will be large, and memory consumption will be high.
+        If `return_coupling_mats` is False, returns `(gw_dmat, None)`.
+        This argument is independent of whether the coupling matrices are written to a file;
+        one may return the coupling matrices, write them to file, both, or neither.
+
+    :return: If `return_coupling_mats` is True,
+        returns `( gw_dmat, couplings )`,
+        where gw_dmat is a square matrix whose (i,j) entry is the GW distance
+        between two cells, and `couplings` is a list of tuples (i,j,
+        coupling_mat) where `i,j` are indices corresponding to positions in the list `cells`
+        and `coupling_mat` is a coupling matrix between the two cells.
+        If `return_coupling_mats` is False, returns `(gw_dmat, None)`.
+    """
+    # Combine queries and targets
+    GW_cells = []
+    for A, a in queries:
+        GW_cells.append(GW_cell(A, a))
+    for A, a in targets:
+        GW_cells.append(GW_cell(A, a))
+
+    num_queries = len(queries)
+    num_targets = len(targets)
+    gw_dmat = np.zeros((num_queries, num_targets))
+    if return_coupling_mats is not None:
+        gw_coupling_mats = []
+    total_num_pairs = num_queries * num_targets
+    # Prepare indices by offsetting the target indices by the number of queries
+    ij = tqdm(it.product(np.arange(num_queries), np.arange(num_targets) + num_queries), total=total_num_pairs)
+    with Pool(
+        initializer=_init_gw_pool, initargs=(GW_cells, ), processes=num_processes
+    ) as pool:
+        gw_data: Iterator[tuple[int, int, Matrix, float]]
+        gw_data = pool.imap_unordered(_gw_index, ij, chunksize=20)
+        if (gw_dist_csv is not None) or (gw_coupling_mat_csv is not None):
+            if names is None:
+                raise Exception(
+                    "Must supply list of cell identifiers for writing to file."
+                )
+            gw_data = csv_output_writer(
```

Comment:
It's nice that you were able to make use of this existing function. When I wrote it I had square matrices in mind but I think this should be correct.